src/Esprima/Ast/SyntaxElement.cs

```diff
@@ -0,0 +1,34 @@
+﻿using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using Esprima.Utils;
+
+namespace Esprima.Ast;
+
+[DebuggerDisplay($"{{{nameof(GetDebuggerDisplay)}(), nq}}")]
+public abstract class SyntaxElement
+{
+    private protected AdditionalDataContainer _additionalDataContainer;
+
+    /// <summary>
+    /// Gets additional, user-defined data associated with the specified key.
+    /// </summary>
+    /// <remarks>
+    /// The operation is not guaranteed to be thread-safe. In case concurrent access or update is possible, the necessary synchronization is caller's responsibility.
+    /// </remarks>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public object? GetAdditionalData(object key) => _additionalDataContainer.GetData(key);
+
+    /// <summary>
+    /// Sets additional, user-defined data associated with the specified key.
+    /// </summary>
+    /// <remarks>
+    /// The operation is not guaranteed to be thread-safe. In case concurrent access or update is possible, the necessary synchronization is caller's responsibility.
+    /// </remarks>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public void SetAdditionalData(object key, object? value) => _additionalDataContainer.SetData(key, value);
+
+    public Range Range;
```

Comment:
these might be candidates for `ref readonly` returns

Reply:
We've already [discussed this](https://github.com/sebastienros/esprima-dotnet/pull/273#issue-1268699911). AFAICS, we can't really do any better without massive breaking changes.

Reply:
Well you have a good memory, unlike I. Location is usually only used on error conditions so doesn't matter that much from Jint's perspective.